### DISCUSSION TITLE
Using Tailwind in a library

### DISCUSSION THREAD

**Discussion by adamzerner** [tailwindlabs/tailwindcss#12594] [Help]
I am in the process of working on a component library called [RFUI](https://rfui.deno.dev/). The idea is that users could do something like:

```
<Button variant="secondary">Submit</Button>
```
and get a nice looking button.

But I'm using Tailwind, so the [code](https://github.com/adamzerner/rfui/blob/master/components/atoms/button.tsx) for my `Button` component has a bunch of Tailwind CSS classes. Something like:

```
const Button = () => {
  return (
    <button class="p-5">{children}</button>
  );
}
```
So then, if someone uses my `Button` component but isn't using Tailwind, it won't work. They won't have the corresponding CSS for `p-5`:

```
.p-5 {
  padding: 20px;
}
```
I see various paths forward here, but I'm wondering if there are any sort of general recommendations or advice for people in situations like this.

---

**1) Ditch Tailwind and use `style`**

So like instead of `<button class="p-5">` it'd be `<button style={{ padding: 20 }}>`. This would make things convenient for users of the library, but inconvenient for me as the author.

It also wouldn't allow for any sort of customization. Like if users of the library want a different "palatte" of spacing values.

**2) Transpile my Tailwind classes to `style`s**

When coding locally, I could write `<button class="p-5">`. Which is convenient for me as the author. And then I could run some sort of script that transpiles it to `<button style={{ padding: 20 }}>` for others to use.

This has the benefit of making things convenient for me as the author. But I don't love the added complexity that the transpilation step adds. More moving pieces. More room for things to go wrong. More stuff to understand.

It also doesn't address the fact that users can't customize things with eg. different spacing values. Which probably isn't a huge deal, but it's something.

**3) Make Tailwind a dependency**

Like, "if you want to use RFUI, you have to also use Tailwind". Not a terrible option, but something I would prefer to avoid.

**4) Make a CSS file a dependency**

I could just give them a CSS file that has styles for the CSS classes my library uses.

```
/* rfui.css */

.p-5 {
  padding: 20px;
}
```

This seems to me like it's the best option by a pretty wide margin. The only downside really is that it adds some amount of complexity vs being able to import components and have them "just work". But the complexity it adds is pretty minimal, I think.

It [looks like](https://www.npmjs.com/package/@shopify/polaris#usage) this is what `@shopify/polaris` does.

As for customization of things like spacing and colors, I could probably just use CSS variables in this file and have them edit those variables if they want.

**Comment by simpel** [comment]
@adamzerner Hi! I dont have a solution I'm afraid but I have exactly the same problem, yey! How did you solve it in the end?

**Reply by adamzerner** [reply]
I went with option 4. Seems like the tradeoffs come out in it's favor.

Relevant docs:
- [Getting started](https://rfui.deno.dev/getting-started)
- [Stylesheet](https://rfui.deno.dev/stylesheet)
- [CSS varaibles](https://rfui.deno.dev/css-variables)

**Reply by ghost** [reply]
How are you preventing the global tailwind css from affecting the parent?

**Reply by ghost** [reply]
Unless I remove `@tailwind base;`, I break the existing styling of my parent. Even the prefix doesn't help.

**Reply by adamzerner** [reply]
@User00701 I'm having trouble understanding what you mean. In particular, I'm not sure what you mean by "parent". Can you clarify and elaborate?

**Reply by ghost** [reply]
Sorry for the misunderstanding, from the parent I meant the host, the consumer for the library. I was finally able to fix it with postcss autoprefix plugin, Basically namespacing all styles in a class.

**Comment by simpel** [comment]
Hey!Yea, we’re leaning towards something similar. Do you optimize the dependency CSS somehow or does it contain all the CSS for every component? Nice lib also! Keep it up!On 16 Feb 2024, at 19:47, Adam Zerner ***@***.***> wrote:﻿
I went with option 4. Seems like the tradeoffs come out in it's favor.
Relevant docs:

Getting started
Stylesheet
CSS varaibles

**Reply by adamzerner** [reply]
> Do you optimize the dependency CSS somehow or does it contain all the CSS for every component?

I minify it. I [_think_](https://github.com/adamzerner/rfui/issues/176#issuecomment-1913708688) I'm also compressing it, but I'm not entirely sure. The file does contain CSS for every component though.

But it's [not](https://rfui.deno.dev/target-audience#page-load-speed) a strong design goal of mine to optimize here very much. For 4G and better network connections the 50 KB shouldn't slow things down more than 0.02 seconds or so.

> Nice lib also! Keep it up!

Thanks!

**Comment by fivaz3** [comment]
depending on your framework you can use unocss with the present for tailwind so it will automatically convert your tailwind classes into a scoped css inside the component so your lib will just works without any adicional css import
